opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -1113,20 +1150,42 @@ const Dune::FieldVector<double,3> CpGrid::faceCenterEcl(int cell_index, int face
     //   0---1
 
     // this follows the DUNE reference cube
-    static const int faceVxMap[ 6 ][ 4 ] = { {0, 2, 4, 6}, // face 0
-                                             {1, 3, 5, 7}, // face 1
-                                             {0, 1, 4, 5}, // face 2
-                                             {2, 3, 6, 7}, // face 3
-                                             {0, 1, 2, 3}, // face 4
-                                             {4, 5, 6, 7}  // face 5
+    static const int faceVxMap[ 6 ][ 4 ] = { {0, 2, 4, 6}, // face 0 - I_FACE false
+                                             {1, 3, 5, 7}, // face 1 - I_FACE true
+                                             {0, 1, 4, 5}, // face 2 - J_FACE false
+                                             {2, 3, 6, 7}, // face 3 - J_FACE true
+                                             {0, 1, 2, 3}, // face 4 - K_FACE false
+                                             {4, 5, 6, 7}  // face 5 - K_FACE true
     };
 
 
     assert (current_view_data_->cell_to_point_[cell_index].size() == 8);
     Dune::FieldVector<double,3> center(0.0);
-    for( int i=0; i<4; ++i )
-    {
-        center += vertexPosition(current_view_data_->cell_to_point_[cell_index][ faceVxMap[ face ][ i ] ]);
+
+    bool isCoarseCellInside = (intersection.inside().level() == 0);
+    bool isCoarseCellOutside = false;
+    if (intersection.neighbor()){
+        isCoarseCellOutside = (intersection.outside().level() == 0);
+    }
+    bool twoCoarseNeighboringCells = isCoarseCellInside && isCoarseCellOutside;
+    bool isOnGridBoundary_coarseNeighboringCell = intersection.boundary() && isCoarseCellInside && (!intersection.neighbor());
+
+    // For CpGrid with LGRs, a refined face with a coarse neighboring cell and a refined neighboring cell
+    // (that is when the face belongs to the boundary of an LGR and is located in the interior of the grid),
+    // unfortunately leads us to a different order of the faces, in cell_to_face_, depending on if the
+    // neighboring cell, here with cell_index index, is the coarse one or the refined one. Preceisely,
+    // cell_to_face_[cell_index - coarse neighboring cell] = { left, right, front, back, bottom, top} = {0,1,2,3,4,5} with
+    // the notation above, and
+    // cell_to_face_[cell_index - refined neighboring cell] = {bottom, front, left, right, back, top} = {2,3,1,4,0,5} with
+    // the notation used in faceVxMap.
+
+    for( int i=0; i<4; ++i ) {
+        if ((maxLevel() == 0) || twoCoarseNeighboringCells || isOnGridBoundary_coarseNeighboringCell) {
+            center += vertexPosition(current_view_data_->cell_to_point_[cell_index][ faceVxMap[ face ][ i ] ]);
+        }
+        else { //  (refined) intersection with one coarse neighboring cell and one refined neighboring cell
+            center += vertexPosition(current_view_data_->face_to_point_[intersection.id()][i]);
+        }
```

Comment:
```suggestion
    if ((maxLevel() == 0) || twoCoarseNeighboringCells || isOnGridBoundary_coarseNeighboringCell) {
        for( int i=0; i<4; ++i ) {
            center += vertexPosition(current_view_data_->cell_to_point_[cell_index][ faceVxMap[ face ][ i ] ]);
        }
    } else { //  (refined) intersection with one coarse neighboring cell and one refined neighboring cell
        center = intersection.geometry().center();
    }
```
I guess the reason for the if-else branches is to be sure to use the old way to prevent different rounding errors.
But for the case where one cell is from the lgr: Is it possible to use the geometry of the cell to get the center?

Just a thought and  a suggestion. It might not even work.
The current code is correct. Even `Intersection.id()` returns the correct index and not an id (a bit confusing, though).

Reply:
I did such a change directly in the file on opm-simulators, (intersection.geometry().center()) and it was not working back then. It could well be that the problem was something else. I would skip this suggestion for now, since, as you mentioned, this version works, even if it's not the most elegant. 

Is there any problem with intersection.id() ? Could it be that it returns the index of the intersection since, in the test for LGRs, all the cells are active? Should I replace intersection.id() with something else? How do we get the actual index of the face if it's not via id() ?